File: SongRanker.cc
#include "Song.h"
#include "Node.h"
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>


int main(int argc, char** argv) {
  // Song* songs = new Song[5];
  // songs[0] = Song("Better than Revenge", "Speak Now", "TS");
  // songs[1] = Song("Lavender Haze", "Midnights", "TS");
  // songs[2] = Song("Paper Rings", "Lover", "TS");
  // songs[3] = Song("Mr. Perfectly Fine", "Fearless", "TS");
  // songs[4] = Song("Welcome to New York", "1989", "TS");
  Song* songs = new Song[190];
  
  std::fstream fin;
  fin.open("TSSongs.csv", std::ios::in);
  if (fin.is_open()) {
    std::string line;
    std::string title, album;
    int j = 0;
    while (getline(fin, line)) {
      std::stringstream ss(line);
      getline(ss, title, ',');
      getline(ss, album, '\r');
      songs[j] = Song(title, album, "Taylor Swift");
      // std::cout << "Added song: " << songs[j] << std::endl;
      j += 1;
    }

    fin.close();
  } else {
    return EXIT_FAILURE;
  }

  // std::cout << "-----------------------" << std::endl;

  std::cout << "Warning: Answering any of the following questions will overwrite the rankings in output.txt" << std::endl;
  std::cout << "For the following questions, answer 1 or 2 (0 or a letter to exit)." << std::endl;
  Node* first = new Node(songs[0]);
  bool result;
  int i;
  for (i = 1; i < 190; i++) {
    result = first->insert(songs[i]);
    if (!result) {
      break;
    }
  }
  if (i > 1) {
    printRankings(*first);
  }
}

File: songranker.py
import csv
import avltree
import random
import datetime

songlist = []
unknown = []
tree = avltree.AVLTree()
csvPath = "songs.csv"
currentTime = datetime.datetime.now().strftime("%m%d%Y%H%M%S")
outputPath = "TSrankings" + currentTime + ".txt"

def listRankings(input):
    output = ""
    for i, song in enumerate(input, start=1):
        output += "\n" + str(i) + ". " + str(song)
    return output

def writeRankings(treeRankings, listRankings, filePath):
    with open(filePath, "w") as f:
        f.write("Song Rankings:")
        f.write(treeRankings)
        f.write("\n-----------------------\nSongs you don't know:")
        f.write(listRankings)

with open(csvPath, newline='') as f:
    reader = csv.reader(f, delimiter = ',')
    for row in reader:
        songlist.append(avltree.Song(row[0], row[1], "Taylor Swift"))
random.shuffle(songlist)

start = 0
for i, song in enumerate(songlist):
    print("Do you know this song? (1 for yes, anything else for no): " + str(song))
    knownAnswer = int(input())
    if knownAnswer != 1:
        unknown.append(song)
        start += 1
    else:
        print("------------")
        print("For the following questions, answer 1 or 2 (0 to exit, 3 for don't know song on the right)")
        tree.insert(song)
        break
    if start % 10 == 0:
        writeRankings(tree.treeRankings(), listRankings(unknown), outputPath)

for i, song in enumerate(songlist[start+1:]):
    result = tree.insert(song)
    if result == 0:
        break
    if result == 3:
        unknown.append(song)
    if i % 10 == 0:
        writeRankings(tree.treeRankings(), listRankings(unknown), outputPath)

writeRankings(tree.treeRankings(), listRankings(unknown), outputPath)